src/vs/vscode.d.ts

```diff
@@ -8940,6 +8940,289 @@ declare module 'vscode' {
+		 */
+		constructor(id: string, resource: Uri, range: Range, comments: Comment[]);
+
+		/**
+		 * Dispose this comment thread.
+		 */
+		dispose(): void;
+	}
+
+	/**
+	 * A comment is displayed within the editor or the Comments Panel, depending on how it is provided.
+	 */
+	export class Comment {
+		/**
+		 * The id of the comment
+		 */
+		readonly id: string;
+
+		/**
+		 * The human-readable comment body
+		 */
+		readonly body: MarkdownString;
+
+		/**
+		 * The display name of the user who created the comment
+		 */
+		readonly userName: string;
+
+		/**
+		 * Optional label describing the [Comment](#Comment)
+		 * Label will be rendered next to userName if exists.
+		 */
+		readonly label?: string;
+
+		/**
+		 * The icon path for the user who created the comment
+		 */
+		readonly userIconPath?: Uri;
+
+		/**
+		 * The command to be executed if the comment is selected in the Comments Panel
+		 */
+		readonly selectCommand?: Command;
+
+		/**
+		 * The command to be executed when users try to save the edits to the comment
+		 */
+		readonly editCommand?: Command;
+
+		/**
+		 * The command to be executed when users try to delete the comment
+		 */
+		readonly deleteCommand?: Command;
+
+		/**
+		 * @param id The id of the comment
+		 * @param body The human-readable comment body
+		 * @param userName The display name of the user who created the comment
+		 */
+		constructor(id: string, body: MarkdownString, userName: string);
+	}
+
+	/**
+	 * The comment input box in Comment Widget.
+	 */
+	export interface CommentInputBox {
+		/**
+		 * Setter and getter for the contents of the comment input box
+		 */
+		value: string;
+
+		/**
+		 * The uri of the document comment input box has been created on
+		 */
+		resource: Uri;
+
+		/**
+		 * The range the comment input box is located within the document
+		 */
+		range: Range;
+	}
+
+	/**
+	 * Commenting range provider for a [comment controller](#CommentController).
+	 */
+	export interface CommentingRangeProvider {
+		/**
+		 * Provide a list of ranges which allow new comment threads creation or null for a given document
+		 */
+		provideCommentingRanges(document: TextDocument, token: CancellationToken): ProviderResult<Range[]>;
+	}
+
+	/**
+	 * Comment thread template for new comment thread creation.
+	 */
+	export interface CommentThreadTemplate {
+		/**
+		 * The human-readable label describing the [Comment Thread](#CommentThread)
+		 */
+		label: string;
+
+		/**
+		 * Optional accept input command
+		 *
+		 * `acceptInputCommand` is the default action rendered on Comment Widget, which is always placed rightmost.
+		 * This command will be invoked when users the user accepts the value in the comment editor.
+		 * This command will disabled when the comment editor is empty.
+		 */
+		acceptInputCommand?: Command;
+
+		/**
+		 * Optional additonal commands.
+		 *
+		 * `additionalCommands` are the secondary actions rendered on Comment Widget.
+		 */
+		additionalCommands?: Command[];
```

Comment:
Wondering if we should work more with menus or not, e.g SCM and TreeView are heavily menu based, e.g you would specify a new menu like `editor/commentInput` to which extensions statically register. The `id` would be exposed as context key to support enable/disablement.